semgrep-core/src/core/Pretty_print_AST.ml

```diff
@@ -770,6 +770,23 @@ and pattern env = function
   | PatId (id, _id_info) -> ident id
   | x -> todo (P x)
 
+let ctype = function
+  | G.Cbool -> "bool"
+  | G.Cint -> "int"
+  | G.Cstr -> "str"
+  | G.Cany -> "???"
+
+let constness env = function
+  | G.NotCst -> "TOP"
+  | G.Sym e -> Printf.sprintf "sym(%s)" (expr env e)
```

Comment:
Just to check: this means `constness` can contain non-constant expressions now, right? Should we rename it?

Reply:
> I would be interested in understanding the cycle issue. I looked at your PR to make constant propagation a true must analysis, but didn't get all the details.

There is a test case to detect if it happens again, I can comment it a bit more. But it's just a for-loop:

```go
for true {
     cpuids = append(cpuids)
}
```

In the original example `cpuids` was declared as `var cupids []int` which I presume initializes it as an empty array. But the analysis didn't catch that and it was considered simply uninitialized. And at the moment the analysis started with the assumption that all locals were `_|_` at entry point. So the analysis inferred that at the beginning of the loop `cpuids` was `append(cpuids)`, because in one branch we have `_|_` and in another we have `append(cpuids)`. So we set the `id_constness` of the second occurrence of `cpuids` to `append(cpuids)`, and that is a cycle (both `cpuids` are the exact same occurrence, so they share the same `id_constness` `ref`).

Of course this wouldn't happen if the analysis knew that `cpuids` was actually initialized to `[]`, but it can always happen something like that.

After #4410 we are actually a bit too strict and initialize all mappings so that variables are assumed non-constant (`T`). But this removes the problem with loops, `T` and `append(cpuids)` gives us `T`.